### PR TITLE
graphql: Add inigo engine fingerprint

### PR DESCRIPTION
**Pull request by kingthorin**
## Overview
- CHANGELOG > Updated note.
- GraphQlFingerprinter > Add new engine handling.
- GraphQlFingerprinterUnitTest > Add coverage for inigo.
- Messages.properties > Add supporting key/value pairs.
- alerts.html > Add supported engine list.

## Related Issues
NA

## Checklist
- [na] Update help
- [x] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [x] Write tests
- [x] Check code coverage
- [x] Sign-off commits
- [x] Squash commits
- [x] Use a descriptive title

### CODE REVIEW

**Comment by psiinon** [comment]
Should we have a help page which lists the engines we can fingerprint?
I couldnt see an existing one.
It doesnt have to be anything cmplicated, just a list, maybe with links to the relevant sites for the engines?

**Comment by psiinon** [comment]
Logo
**Checkmarx One – Scan Summary & Details** – [31732a81\-d522\-4665\-8302\-104ab4aa546a](https://eu.ast.checkmarx.net/projects/0da811fa-6373-42b1-bbb6-d0026285b136/scans?id=31732a81-d522-4665-8302-104ab4aa546a&branch=kingthorin%2Fgql-inigo)
### No New Or Fixed Issues Found

**Comment by kingthorin** [comment]
Just seems like a pain to have to maintain.

Alerts (if any) have the details.

**Comment by psiinon** [comment]
I disagree. I think its very useful for users to be able to see somewhere which engines we can fingerprint.
Its not like there are very many, or that they change frequently. 
And then theres more chances of people telling us which ones we're missing?

**Comment by kingthorin** [comment]
Okay. Want that added in this PR?

**Comment by psiinon** [comment]
Please :) Can just be a list added to a new section on this help page https://www.zaproxy.org/docs/desktop/addons/graphql-support/alerts/

**Comment by kingthorin** [comment]
Added.,

**Comment by kingthorin** [comment]
For future reference I used a regex like: `graphql\.engine.*docsUrl = (.*)\ngraphql\.engine.*name = (.*)\n.*\n\n` then replaced with `$2 - $1` then manipulated it in Notepad++ for the `li` etc (that could also be done in the replace value).

**Review by psiinon — APPROVED** [review]

File: addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlFingerprinter.java [hunk 1]
```diff
@@ -459,6 +460,22 @@ private boolean checkHyperGraphQlEngine() {
                 "Validation error of type UnknownDirective: Unknown directive deprecated @ '__typename'");
     }
 
+    private boolean checkInigoEngine() {
```
Comment (thc202) [inline addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlFingerprinter.java:463]: Where's the source link?

File: addOns/graphql/src/main/javahelp/org/zaproxy/addon/graphql/resources/help/contents/alerts.html [hunk 2]
```diff
@@ -29,6 +29,47 @@ <h1 id="id-50007">GraphQL Alerts</h1>
...
+  <li>Absinthe - https://github.com/absinthe-graphql/absinthe</li>
+  <li>Agoo - https://github.com/ohler55/agoo</li>
+  <li>Apollo - https://github.com/apollographql/apollo-server</li>
+  <li>Ariadne - https://github.com/mirumee/ariadne</li>
+  <li>AWS AppSync - https://aws.amazon.com/appsync</li>
+  <li>caliban - https://github.com/ghostdogpr/caliban</li>
+  <li>Dgraph - https://github.com/dgraph-io/dgraph</li>
+  <li>Diana.jl - https://github.com/neomatrixcode/Diana.jl</li>
+  <li>Directus - https://github.com/directus/directus</li>
+  <li>Example GraphQL Engine - https://example.com/graphql-engine-reference</li>
```
Comment (thc202) [inline addOns/graphql/src/main/javahelp/org/zaproxy/addon/graphql/resources/help/contents/alerts.html:44]: This does not look right.

File: addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlFingerprinter.java (hunk 1 above)
Comment (kingthorin) [reply addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlFingerprinter.java:463]: The project doesn't see to be open source. Do you want me to link/comment the graphw00f PR?

Comment (kingthorin) [reply addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlFingerprinter.java:463]: Added

File: addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlFingerprinter.java [hunk 3]
```diff
@@ -459,6 +460,23 @@ private boolean checkHyperGraphQlEngine() {
...
+        // https://github.com/dolevf/graphw00f/commit/52e25d376f5fd4dcad062ba79a1b6c3e5e1c68dc
+        sendQuery("query {__typename}");
+        if (lastQueryMsg != null && lastQueryMsg.getResponseHeader().isJson()) {
+            try {
+                String response = lastQueryMsg.getResponseBody().toString();
+                JsonNode inigo = OBJECT_MAPPER.readValue(response, JsonNode.class).get("inigo");
```
[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlFingerprinter.java:469]: You're right I totally interpreted it backwards

[the remark this answers is not in the document]
Comment (thc202) [reply addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlFingerprinter.java]: I'd remove the object check, we don't know the actual type (from the referenced code) and based on the plural I think it's more likely to be an array than an object.

[the remark this answers is not in the document]
Comment (thc202) [reply addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlFingerprinter.java]: Or not, it's probably an object based on https://github.com/inigolabs/fezzik/blob/3650ea733e4e65a48bb3a4b1b092228d67362676/client/error.go#L15

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**